jupyter/swan.py

```diff
@@ -0,0 +1,560 @@
+        return os.path.exists(self._filename(experiment_id))
+
+    def __setitem__(self, experiment_id, df):
+        if not os.path.exists(self.CACHE_DIR):
+            os.makedirs(self.CACHE_DIR)
+        df.to_csv(self._filename(experiment_id), compression='bz2')
+
+    def __getitem__(self, experiment_id):
+        if experiment_id in self:
+            return pd.read_csv(self._filename(experiment_id), compression='bz2')
+        else:
+            raise KeyError()
+
+    def __call__(self, fun):
+        """ Can be use as decorator for function that have experiment_id as first parameter and returns dataframe."""
+        def _dec(experiment_id, *args, **kw):
+            if experiment_id in self:
+                return self[experiment_id]
+
+            df = fun(experiment_id, *args, **kw)
+
+            self[experiment_id] = df
+            return df
+        return _dec
+
+
+# -------------------------------
+# cassandra session singleton
+# -------------------------------
+CASSANDRA_SESSION = None  # one instance for all existing notebook experiments
+DEFAULT_CASSANDRA_OPTIONS = dict(
+    nodes=['127.0.0.1'],
+    port=9042,
+    ssl_options=None
+)
+
+
+def _get_or_create_cassandra_session(nodes, port, ssl_options=None):
+    """ Get or prepare new session to Cassandra cluster.
+
+    :param nodes: List of addresses of cassandra nodes.
+    :param port: Port of cassandra service listening on.
+    :param ssl_options: Optional SSL options to connect to cassandra in secure manner.
+    :returns: cassandra session singleton
+
+    """
+    from cassandra.auth import PlainTextAuthProvider
+    from cassandra.cluster import Cluster
+    from cassandra.query import ordered_dict_factory
+    global CASSANDRA_SESSION
+    if not CASSANDRA_SESSION:
+        auth_provider = None
+        if ssl_options:
+            username = ssl_options.pop('username')
+            password = ssl_options.pop('password')
+            auth_provider = PlainTextAuthProvider(username, password)
+
+        cluster = Cluster(nodes, port=port, ssl_options=ssl_options, auth_provider=auth_provider)
+        CASSANDRA_SESSION = cluster.connect()
+        CASSANDRA_SESSION.row_factory = ordered_dict_factory
+    return CASSANDRA_SESSION
+
+
+# ----------------------------------------
+# helper function to load and convert data
+# ----------------------------------------
+
+
+def _load_rows_from_cassandra(experiment_id, cassandra_session, keyspace='snap'):
+    """ Load data from cassandra database as rows.
+
+    It will cache loaded rows from cassandra to speed up further requests.
+    It only loads doubleval, ns and tags values ignoring other kinds of types.
+
+    :returns: Data as cassandra rows (each row being dict like).
+    """
+
+    from cassandra.query import SimpleStatement
+
+    query = """SELECT ns, doubleval, tags
+               FROM %s.metrics
+               WHERE tags['swan_experiment'] = \'%s\'  ALLOW FILTERING""" % (
```

Comment:
good catch but unfortunately, cannot keyspace use 

```
SyntaxException: <Error from server: code=2000 [Syntax error in CQL query] message="Bind variables cannot be used for keyspace names">
``` 

but we good (and fixed) for experiment_id which is explicitly required by user !